api/tests/unit/app/test_unit_app_routers.py

```diff
@@ -0,0 +1,188 @@
+
+    # When
+    result = router.db_for_read(FFAdminUser)
+
+    # Then
+    # Read strategy DISTRIBUTED is random, so just this is a check
+    # against loading the primary or one of the cross region replicas
+    assert result.startswith("cross_region_replica_")
+
+    # Check that the number of replica call counts is as expected.
+    conn_call_count = 6
+    assert create_connection_patch.call_count == conn_call_count
+    assert conn_patch.is_usable.call_count == conn_call_count
+
+
+def test_replica_router_db_for_read_with_all_offline_replicas(
+    settings: SettingsWrapper,
+    mocker: MockerFixture,
+    reset_cache: None,
+) -> None:
+    # Given
+    settings.NUM_DB_REPLICAS = 4
+    settings.NUM_CROSS_REGION_DB_REPLICAS = 2
+    settings.REPLICA_READ_STRATEGY = "DISTRIBUTED"
+
+    conn_patch = mocker.MagicMock()
+
+    # All replicas go offline.
+    conn_patch.is_usable.return_value = False
+    create_connection_patch = mocker.patch(
+        "app.routers.connections.create_connection", return_value=conn_patch
+    )
+
+    router = PrimaryReplicaRouter()
+
+    # When
+    result = router.db_for_read(FFAdminUser)
+
+    # Then
+    # Fallback to primary database if all replicas are offline.
+    assert result == "default"
+
+    # Check that the number of replica call counts is as expected.
+    conn_call_count = 6
+    assert create_connection_patch.call_count == conn_call_count
+    assert conn_patch.is_usable.call_count == conn_call_count
+
+
+def test_replica_router_db_with_sequential_read(
+    settings: SettingsWrapper,
+    mocker: MockerFixture,
+    reset_cache: None,
+) -> None:
+    # Given
+    settings.NUM_DB_REPLICAS = 100
+    settings.NUM_CROSS_REGION_DB_REPLICAS = 2
+    settings.REPLICA_READ_STRATEGY = "SEQUENTIAL"
+
+    conn_patch = mocker.MagicMock()
+
+    # First replica is offline, so must fall back to second one.
+    conn_patch.is_usable.side_effect = (False, True)
+    create_connection_patch = mocker.patch(
+        "app.routers.connections.create_connection", return_value=conn_patch
+    )
+
+    router = PrimaryReplicaRouter()
+
+    # When
+    result = router.db_for_read(FFAdminUser)
+
+    # Then
+    # Fallback from first replica to second one.
+    assert result == "replica_2"
+
+    # Check that the number of replica call counts is as expected.
+    conn_call_count = 2
+    assert create_connection_patch.call_count == conn_call_count
+    assert conn_patch.is_usable.call_count == conn_call_count
+
+
+def test_replica_router_db_no_replicas(
+    settings: SettingsWrapper,
+    mocker: MockerFixture,
+    reset_cache: None,
+) -> None:
+    # Given
+    settings.NUM_DB_REPLICAS = 0
+    settings.NUM_CROSS_REGION_DB_REPLICAS = 0
+
+    conn_patch = mocker.MagicMock()
+
+    # All replicas should be ignored.
+    conn_patch.is_usable.return_value = True
+    create_connection_patch = mocker.patch(
+        "app.routers.connections.create_connection", return_value=conn_patch
+    )
```

Comment:
Do we actually need this then?

Reply:
For this one I wanted to make sure that the test would fail if there were replicas present and that all the call counts to the `create_connection_patch` and to the `conn_patch.is_usable` mocks were zero. May be a bit overkill, but I think it's better than introducing a test which would silently fail by falling back to the default database after already trying the replicas and failing there.

Reply:
I'd argue that at least this code is redundant then though, right? 

```
conn_patch.is_usable.return_value = True
```